File: sdk/examples/python/nis2_assessment.py
"""
nis2_assessment.py — NIS2 Compass assessment workflow using the opensecstack Python SDK.

Steps performed:
  1. Create a NIS2CompassClient from environment variables.
  2. Register a new organisation.
  3. Create a NIS2 assessment under that organisation.
  4. Stream the SARIF report to a local file.

Usage
-----
  # Install the SDK (from the sdk/python directory):
  pip install -e sdk/python

  # Set required environment variables:
  export NIS2_URL=https://nis2.example.com
  export NIS2_API_KEY=<your-api-key>

  # Optional overrides:
  export NIS2_ORG_NAME="Example Organisation"
  export NIS2_ORG_COUNTRY=DE

  python sdk/examples/python/nis2_assessment.py
"""

from __future__ import annotations

import logging
import os
import sys
from pathlib import Path
from opensecstack.nis2compass import NIS2CompassClient
from opensecstack.exceptions import APIError

logging.basicConfig(
    level=logging.INFO,
    format="%(asctime)s %(levelname)s %(message)s",
)
log = logging.getLogger(__name__)

# -------------------------------------------------------------------------
# Helpers
# -------------------------------------------------------------------------


def require_env(key: str) -> str:
    """Return the value of *key* or exit with a descriptive error."""
    value = os.environ.get(key, "").strip()
    if not value:
        sys.exit(f"ERROR: required environment variable {key!r} is not set")
    return value


def env_or_default(key: str, default: str) -> str:
    """Return the value of *key* or *default* when unset/empty."""
    return os.environ.get(key, "").strip() or default


# -------------------------------------------------------------------------
# Main
# -------------------------------------------------------------------------


def main() -> None:
    # ------------------------------------------------------------------
    # 1. Build the NIS2 Compass client.
    # ------------------------------------------------------------------
    base_url = require_env("NIS2_URL")
    api_key = require_env("NIS2_API_KEY")

    client = NIS2CompassClient(base_url=base_url, api_key=api_key, timeout=30)

    # ------------------------------------------------------------------
    # 2. Register a new organisation.
    # ------------------------------------------------------------------
    org_name = env_or_default("NIS2_ORG_NAME", "Example Organisation")
    org_country = env_or_default("NIS2_ORG_COUNTRY", "DE")

    log.info("Creating organisation: name=%r country=%s", org_name, org_country)
    org = client.create_organisation(
        name=org_name,
        industry="Technology",
        country=org_country,
        size="medium",
        entity_type="important",
    )
    org_id: str = org["id"]
    log.info("Organisation created: id=%s name=%s", org_id, org.get("name"))

    # ------------------------------------------------------------------
    # 3. Create a NIS2 assessment.
    #
    # The server automatically seeds 10 control entries (NIS2 Article 21(2)
    # measures a–j) when a new assessment is created.
    # ------------------------------------------------------------------
    log.info("Creating NIS2 assessment for org %s …", org_id)
    assessment = client.create_assessment(
        org_id=org_id,
        title="Initial NIS2 Compliance Assessment",
        framework_version="NIS2-2022/0383",
        scope="All production systems and supporting infrastructure",
        assessor="Security Team",
    )
    assessment_id: str = assessment["id"]
    log.info(
        "Assessment created: id=%s status=%s",
        assessment_id,
        assessment.get("status"),
    )

    # ------------------------------------------------------------------
    # 4. Stream the SARIF report to a local file.
    #
    # SARIF is the recommended format for integrating NIS2 gap findings
    # into code-review pipelines and GitHub Advanced Security.
    # stream_report() writes the response body in chunks so large reports
    # never consume unbounded memory.
    # ------------------------------------------------------------------
    report_path = Path(f"nis2-assessment-{assessment_id}.sarif")
    log.info("Streaming SARIF report to %s …", report_path)
    with report_path.open("wb") as fh:
        try:
            client.stream_report(
                assessment_id=assessment_id,
                format="sarif",
                dest=fh,
            )
        except APIError as exc:
            sys.exit(f"Failed to retrieve SARIF report: {exc}")

    log.info(
        "SARIF report written to %s (%d bytes)",
        report_path,
        report_path.stat().st_size,
    )

    # ------------------------------------------------------------------
    # Summary
    # ------------------------------------------------------------------
    print("\n--- Summary ---")
    print(f"  Organisation : {org.get('name')} ({org_id})")
    print(f"  Assessment   : {assessment.get('title')} ({assessment_id})")
    print(f"  SARIF report : {report_path}")
    print("\nTip: call delete_assessment / delete_organisation to clean up test data.")


if __name__ == "__main__":
    main()

File: sdk/rust/examples/nis2_assessment.rs
//! # NIS2 Assessment Example
//!
//! Demonstrates the full NIS2 Compass workflow:
//!   1. Create (or reuse) an organisation
//!   2. Create an assessment
//!   3. Retrieve controls and patch each to "partially_compliant"
//!   4. Upload a policy document as an artifact
//!   5. Stream a PDF compliance report to `./compliance_report.pdf`
//!
//! ## Usage
//!
//! ```bash
//! NIS2_URL=https://nis2compass.example.com \
//! NIS2_API_KEY=nk_live_... \
//! cargo run --example nis2_assessment
//! ```

use opensecstack::{
    ArtifactType, ControlStatus, CreateAssessmentRequest, CreateOrganisationRequest,
    NIS2CompassClient, OrganisationSize, PatchAssessmentRequest, PatchControlRequest,
};
use std::env;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let base_url = env::var("NIS2_URL").expect("NIS2_URL must be set");
    let api_key = env::var("NIS2_API_KEY").expect("NIS2_API_KEY must be set");

    let client = NIS2CompassClient::new(&base_url, &api_key);

    // ------------------------------------------------------------------
    // 1. Create organisation
    // ------------------------------------------------------------------
    println!("Creating organisation...");
    let org = client
        .create_organisation(CreateOrganisationRequest {
            name: "Example GmbH".to_string(),
            industry: Some("financial_services".to_string()),
            country: Some("DE".to_string()),
            size: Some(OrganisationSize::Medium),
            registration_number: Some("HRB 123456 B".to_string()),
        })
        .await?;
    println!("Organisation created: {} ({})", org.name, org.id);

    // ------------------------------------------------------------------
    // 2. Create assessment
    // ------------------------------------------------------------------
    println!("\nCreating assessment...");
    let assessment = client
        .create_assessment(
            &org.id.to_string(),
            CreateAssessmentRequest {
                title: "NIS2 Initial Assessment 2024".to_string(),
                scope: Some("All IT systems processing customer data".to_string()),
                assessor: Some("ciso@example.de".to_string()),
                due_date: Some("2024-06-30".to_string()),
            },
        )
        .await?;
    println!(
        "Assessment created: {} [{:?}]",
        assessment.id, assessment.status
    );

    // ------------------------------------------------------------------
    // 3. Get controls and update them
    // ------------------------------------------------------------------
    println!("\nFetching controls...");
    let controls = client.get_controls(&assessment.id.to_string()).await?;
    println!("  {} controls returned", controls.len());

    for control in &controls {
        println!(
            "  Patching control '{}' ({}) — current: {:?}",
            control.measure_ref, control.title, control.status
        );
        let updated = client
            .patch_control(
                &assessment.id.to_string(),
                control.measure_ref,
                PatchControlRequest {
                    status: Some(ControlStatus::PartiallyCompliant),
                    evidence: Some(serde_json::json!({
                        "reviewed": true,
                        "reviewer": "ciso@example.de"
                    })),
                    remediation: Some(
                        "Gap analysis complete. Remediation roadmap in progress.".to_string(),
                    ),
                    risk_score: Some(5.0),
                },
            )
            .await?;
        println!("    Updated to: {:?}", updated.status);
    }

    // ------------------------------------------------------------------
    // 4. Upload a policy document (if a local file exists)
    // ------------------------------------------------------------------
    let policy_path = env::var("POLICY_FILE").unwrap_or_else(|_| "policy.pdf".to_string());
    if tokio::fs::metadata(&policy_path).await.is_ok() {
        println!("\nUploading policy document: {policy_path}");
        let artifact = client
            .upload_artifact(
                &assessment.id.to_string(),
                &policy_path,
                ArtifactType::Policy,
                None, // not tied to a specific control
                Some("Information Security Policy v2.0"),
            )
            .await?;
        println!(
            "Artifact uploaded: {} ({} bytes, hash: {})",
            artifact.filename, artifact.size_bytes, artifact.file_hash
        );
    } else {
        println!("\nSkipping artifact upload (POLICY_FILE not set or file not found).");
    }

    // ------------------------------------------------------------------
    // 5. Move assessment to "in_progress"
    // ------------------------------------------------------------------
    println!("\nUpdating assessment status to in_progress...");
    use opensecstack::AssessmentStatus;
    let updated_assessment = client
        .patch_assessment(
            &assessment.id.to_string(),
            PatchAssessmentRequest {
                status: Some(AssessmentStatus::InProgress),
                ..Default::default()
            },
        )
        .await?;
    println!("Assessment status: {:?}", updated_assessment.status);
    println!(
        "Compliance: {:.1}% ({}/{} controls compliant or partial)",
        updated_assessment.stats.compliance_pct,
        updated_assessment.stats.compliant + updated_assessment.stats.partially_compliant,
        updated_assessment.stats.total
    );

    // ------------------------------------------------------------------
    // 6. Stream PDF report
    // ------------------------------------------------------------------
    let report_path = "compliance_report.pdf";
    println!("\nGenerating PDF report to {report_path}...");
    let mut f = tokio::fs::File::create(report_path).await?;
    client
        .get_report_stream(&assessment.id.to_string(), "pdf", &mut f)
        .await?;
    println!("Report saved to {report_path}");

    // ------------------------------------------------------------------
    // 7. Print recent audit log entries
    // ------------------------------------------------------------------
    println!("\nRecent audit entries:");
    let entries = client.get_audit_log(5, 1).await?;
    for entry in &entries {
        println!(
            "  [{}] {} → {} (chain: {}...)",
            entry.created_at.format("%H:%M:%S"),
            entry.actor_id,
            entry.action,
            &entry.chain_hash[..16.min(entry.chain_hash.len())]
        );
    }

    println!("\nDone.");
    Ok(())
}
